File: Entregas/las-middleware/src/logic/posiciones.js
function calcularPosiciones(grupo, partidos) {
  const equiposGrupo = obtenerEquiposPorGrupo(grupo);
  const stats = {};

  equiposGrupo.forEach(e => {
    stats[e.id] = { pj: 0, pg: 0, pe: 0, pp: 0, gf: 0, gc: 0, dg: 0, pts: 0 };
  });

  const partidosGrupo = partidos.filter(p => p.fase === 'grupos' && p.grupo === grupo && p.resultado);

  partidosGrupo.forEach(p => {
    const { golesLocal, golesVisitante } = p.resultado;
    const local = stats[p.local];
    const visitante = stats[p.visitante];

    if (golesLocal === null || golesVisitante === null) return;

    local.pj++; visitante.pj++;
    local.gf += golesLocal; local.gc += golesVisitante;
    visitante.gf += golesVisitante; visitante.gc += golesLocal;

    if (golesLocal > golesVisitante) {
      local.pg++; local.pts += 3;
      visitante.pp++;
    } else if (golesLocal < golesVisitante) {
      visitante.pg++; visitante.pts += 3;
      local.pp++;
    } else {
      local.pe++; local.pts += 1;
      visitante.pe++; visitante.pts += 1;
    }
  });

  Object.values(stats).forEach(s => { s.dg = s.gf - s.gc; });

  const ordenado = equiposGrupo.map(e => ({
    ...e,
    ...stats[e.id]
  })).sort((a, b) => {
    if (b.pts !== a.pts) return b.pts - a.pts;
    if (b.dg !== a.dg) return b.dg - a.dg;
    if (b.gf !== a.gf) return b.gf - a.gf;

    const enfrentamiento = resolverEnfrentamientoDirecto(a.id, b.id, partidosGrupo);
    if (enfrentamiento !== 0) return enfrentamiento;

    return a.nombre.localeCompare(b.nombre);
  });

  return ordenado;
}

function resolverEnfrentamientoDirecto(idA, idB, partidos) {
  const partido = partidos.find(p =>
    (p.local === idA && p.visitante === idB) ||
    (p.local === idB && p.visitante === idA)
  );

  if (!partido || !partido.resultado) return 0;

  const { golesLocal, golesVisitante } = partido.resultado;
  if (golesLocal === null || golesVisitante === null) return 0;

  let puntosA = 0, puntosB = 0;
  if (partido.local === idA) {
    if (golesLocal > golesVisitante) puntosA = 3;
    else if (golesLocal < golesVisitante) puntosB = 3;
    else { puntosA = 1; puntosB = 1; }
  } else {
    if (golesVisitante > golesLocal) puntosA = 3;
    else if (golesVisitante < golesLocal) puntosB = 3;
    else { puntosA = 1; puntosB = 1; }
  }

  return puntosB - puntosA;
}

function obtenerPrimerosDos(grupo, partidos) {
  const posiciones = calcularPosiciones(grupo, partidos);
  const tieneResultados = partidos.some(p => p.fase === 'grupos' && p.grupo === grupo && p.resultado);
  return {
    primero: tieneResultados && posiciones.length > 0 ? posiciones[0] : null,
    segundo: tieneResultados && posiciones.length > 1 ? posiciones[1] : null,
    tercero: posiciones.length > 2 ? posiciones[2] : null
  };
}

function obtenerTercerosPuestos(partidos) {
  const terceros = [];

  gruposDisponibles.forEach(grupo => {
    const tieneResultados = partidos.some(p => p.fase === 'grupos' && p.grupo === grupo && p.resultado);
    const posiciones = calcularPosiciones(grupo, partidos);
    if (tieneResultados && posiciones.length >= 3) {
      terceros.push({
        ...posiciones[2],
        grupo
      });
    }
  });

  const ordenado = terceros.sort((a, b) => {
    if (b.pts !== a.pts) return b.pts - a.pts;
    if (b.dg !== a.dg) return b.dg - a.dg;
    if (b.gf !== a.gf) return b.gf - a.gf;
    return a.nombre.localeCompare(b.nombre);
  });

  return ordenado;
}

function obtenerClasificados(partidos) {
  const primeros = [];
  const segundos = [];

  gruposDisponibles.forEach(grupo => {
    const tieneResultados = partidos.some(p => p.fase === 'grupos' && p.grupo === grupo && p.resultado);
    const pos = calcularPosiciones(grupo, partidos);
    if (tieneResultados && pos.length >= 1) primeros.push(pos[0]);
    if (tieneResultados && pos.length >= 2) segundos.push(pos[1]);
  });

  const terceros = obtenerTercerosPuestos(partidos);
  const mejoresTerceros = terceros.slice(0, 8);

  return { primeros, segundos, mejoresTerceros, todosTerceros: terceros };
}

File: Entregas/las-middleware/backend/logic/posiciones.py
from sqlalchemy.orm import Session
from models import Match
from schemas import StandingRow, ThirdPlaceRow

GRUPOS = ["A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L"]


def calcular_posiciones(grupo: str, db: Session) -> list[StandingRow]:
    from models import Team

    equipos = db.query(Team).filter(Team.group_letter == grupo).all()
    stats = {e.id: {"pj": 0, "pg": 0, "pe": 0, "pp": 0, "gf": 0, "gc": 0, "dg": 0, "pts": 0} for e in equipos}

    partidos = (
        db.query(Match)
        .filter(Match.group_letter == grupo, Match.phase == "grupos", Match.local_goals != None)
        .all()
    )

    for p in partidos:
        gl, gv = p.local_goals, p.visitor_goals
        loc = stats[p.local_team_id]
        vis = stats[p.visitor_team_id]

        loc["pj"] += 1; vis["pj"] += 1
        loc["gf"] += gl; loc["gc"] += gv
        vis["gf"] += gv; vis["gc"] += gl

        if gl > gv:
            loc["pg"] += 1; loc["pts"] += 3; vis["pp"] += 1
        elif gv > gl:
            vis["pg"] += 1; vis["pts"] += 3; loc["pp"] += 1
        else:
            loc["pe"] += 1; loc["pts"] += 1
            vis["pe"] += 1; vis["pts"] += 1

    for s in stats.values():
        s["dg"] = s["gf"] - s["gc"]

    ordenado: list[StandingRow] = []
    for e in equipos:
        s = stats[e.id]
        ordenado.append(StandingRow(
            id=e.id, name=e.name, flag=e.flag, group_letter=e.group_letter,
            pj=s["pj"], pg=s["pg"], pe=s["pe"], pp=s["pp"],
            gf=s["gf"], gc=s["gc"], dg=s["dg"], pts=s["pts"],
        ))

    ordenado.sort(key=lambda r: (
        -r.pts, -r.dg, -r.gf, r.name
    ))

    return ordenado


def obtener_terceros(db: Session) -> list[ThirdPlaceRow]:
    terceros = []
    for g in GRUPOS:
        pos = calcular_posiciones(g, db)
        if len(pos) >= 3:
            t = pos[2]
            terceros.append(ThirdPlaceRow(
                id=t.id, name=t.name, flag=t.flag, group_letter=t.group_letter,
                pj=t.pj, pg=t.pg, pe=t.pe, pp=t.pp,
                gf=t.gf, gc=t.gc, dg=t.dg, pts=t.pts,
                grupo=g,
            ))

    terceros.sort(key=lambda r: (-r.pts, -r.dg, -r.gf, r.name))
    return terceros


def obtener_clasificados(db: Session) -> tuple[list[StandingRow], list[StandingRow], list[ThirdPlaceRow], list[ThirdPlaceRow]]:
    primeros = []
    segundos = []

    for g in GRUPOS:
        tiene_resultados = db.query(Match).filter(
            Match.group_letter == g, Match.phase == "grupos",
            Match.local_goals != None
        ).count() > 0

        if not tiene_resultados:
            continue
        pos = calcular_posiciones(g, db)
        if len(pos) >= 1:
            primeros.append(pos[0])
        if len(pos) >= 2:
            segundos.append(pos[1])

    todos_terceros = obtener_terceros(db)
    mejores_terceros = todos_terceros[:8]

    return primeros, segundos, mejores_terceros, todos_terceros
